code/datums/action.dm

```diff
@@ -0,0 +1,406 @@
+#define AB_CHECK_RESTRAINED 1
+#define AB_CHECK_STUNNED 2
+#define AB_CHECK_LYING 4
+#define AB_CHECK_CONSCIOUS 8
+
+
+/datum/action
+	var/name = "Generic Action"
+	var/desc = null
+	var/obj/target = null
+	var/check_flags = 0
+	var/processing = 0
+	var/obj/screen/movable/action_button/button = null
+	var/button_icon = 'icons/mob/actions.dmi'
+	var/background_icon_state = "bg_default"
+
+	var/icon_icon = 'icons/mob/actions.dmi'
+	var/button_icon_state = "default"
+	var/mob/owner
+
+/datum/action/New(var/Target)
+	target = Target
+	button = new
+	button.linked_action = src
+	button.name = name
+
+/datum/action/Destroy()
+	if(owner)
+		Remove(owner)
+	if(target)
+		target = null
+	qdel(button)
+	button = null
+	return ..()
+
+/datum/action/proc/Grant(mob/living/L)
```

Comment:
Surprised you didn't cast this to all mobs in case we want ghost action buttons or something.

Ghost boo spell action button, yeahhhh!